File: app/src/main/java/org/cerion/stockcharts/repository/PriceListSQLRepository.java
package org.cerion.stockcharts.repository;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import org.cerion.stockcharts.database.StockDBOpenHelper;
import org.cerion.stocklist.Price;
import org.cerion.stocklist.PriceList;
import org.cerion.stocklist.model.HistoricalDates;
import org.cerion.stocklist.model.Interval;
import org.cerion.stocklist.repository.PriceListRepository;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class PriceListSQLRepository extends SQLiteRepositoryBase implements PriceListRepository {
    private static final String TAG = PriceListSQLRepository.class.getSimpleName();

    public PriceListSQLRepository(Context context) {
        super(StockDBOpenHelper.getInstance(context));
    }

    /*
    public PriceList getLatest(String symbolDescription, Interval interval) throws Exception {
        HistoricalDates dates = getHistoricalDates(symbolDescription, interval);
        boolean update = false;

        if(dates == null) {
            update = true;
        } else {
            Date now = new Date();
            long diff = now.getTime() - dates.LastUpdated.getTime();
            diff /= 1000 * 60 * 60 * 24;
            Log.d(TAG, symbolDescription + " " + interval.name() + " last updated " + dates.LastUpdated + " (" + diff + " days ago)");

            // TODO, smarter updates based on last price obtained and weekends
            if(interval == Interval.DAILY && diff >= 1)
                update = true;
            else if(interval == Interval.WEEKLY && diff > 3)
                update = true;
            else if(interval == Interval.MONTHLY && diff > 7)
                update = true;
        }

        if(update)
            updatePrices(symbolDescription, interval);

        return get(symbolDescription, interval);
    }
    */

    @Override
    public PriceList get(String symbol, Interval interval, int max) {
        throw new RuntimeException("not implemented");
        //List<Price> prices = get(symbolDescription, interval);
        ///return new PriceList(symbolDescription,prices);
    }

    @Override
    public List<Price> get(String symbol, Interval interval) {
        String where = String.format(StockDBOpenHelper.Prices._SYMBOL + "='%s'", symbol);
        Cursor c = db.query(StockDBOpenHelper.Prices.getTableName(interval), null, where, null, null, null, StockDBOpenHelper.Prices._DATE + " ASC");
        List<Price> prices = new ArrayList<>();

        if(c != null) {
            while (c.moveToNext()) {
                Date date = new Date(c.getLong(c.getColumnIndexOrThrow(StockDBOpenHelper.Prices._DATE)));
                float open = c.getFloat(c.getColumnIndexOrThrow(StockDBOpenHelper.Prices._OPEN));
                float high = c.getFloat(c.getColumnIndexOrThrow(StockDBOpenHelper.Prices._HIGH));
                float low = c.getFloat(c.getColumnIndexOrThrow(StockDBOpenHelper.Prices._LOW));
                float close = c.getFloat(c.getColumnIndexOrThrow(StockDBOpenHelper.Prices._CLOSE));
                long volume = c.getLong(c.getColumnIndexOrThrow(StockDBOpenHelper.Prices._VOLUME));

                Price p = new Price(date,open,high,low,close,volume);
                prices.add(p);
            }
            c.close();
        }

        return prices;
    }

    @Override
    public void add(PriceList list) {
        db.beginTransaction();

        String table = StockDBOpenHelper.Prices.getTableName(list.getInterval());
        //Delete current data before re-adding new
        delete(table, String.format("%s='%s'", StockDBOpenHelper.Prices._SYMBOL, list.mSymbol));

        for(Price p : list) {
            ContentValues values = new ContentValues();
            values.put(StockDBOpenHelper.Prices._SYMBOL, list.mSymbol);
            values.put(StockDBOpenHelper.Prices._DATE, p.date.getTime());
            values.put(StockDBOpenHelper.Prices._OPEN, p.open);
            values.put(StockDBOpenHelper.Prices._HIGH, p.high);
            values.put(StockDBOpenHelper.Prices._LOW, p.low);
            values.put(StockDBOpenHelper.Prices._CLOSE, p.getClose());
            values.put(StockDBOpenHelper.Prices._VOLUME, p.volume);
            insert(table, values);
        }

        //Add dates
        // TODO refactor to private function
        Date first = list.getDates()[0];
        Date last = list.getDates()[list.size() - 1];
        table = StockDBOpenHelper.HistoricalDates.getTableName(list.getInterval());
        ContentValues values = new ContentValues();
        values.put(StockDBOpenHelper.HistoricalDates._SYMBOL, list.mSymbol);
        values.put(StockDBOpenHelper.HistoricalDates._FIRST, first.getTime());
        values.put(StockDBOpenHelper.HistoricalDates._LAST, last.getTime());
        values.put(StockDBOpenHelper.HistoricalDates._UPDATED, new Date().getTime());
        db.insertWithOnConflict(table, null, values, SQLiteDatabase.CONFLICT_REPLACE);

        db.setTransactionSuccessful();
        db.endTransaction();
    }

    @Override
    public HistoricalDates getHistoricalDates(String symbol, Interval interval) {
        String where = String.format(StockDBOpenHelper.HistoricalDates._SYMBOL + "='%s'", symbol);
        Cursor c = db.query(StockDBOpenHelper.HistoricalDates.getTableName(interval), null, where, null, null, null, null);
        HistoricalDates result = null;
        if(c != null) {
            if (c.moveToFirst()) {
                result = new HistoricalDates();
                result.Symbol = symbol;
                result.FirstDate = new Date(c.getLong(c.getColumnIndexOrThrow(StockDBOpenHelper.HistoricalDates._FIRST)));
                result.LastDate = new Date(c.getLong(c.getColumnIndexOrThrow(StockDBOpenHelper.HistoricalDates._LAST)));
                result.LastUpdated = new Date(c.getLong(c.getColumnIndexOrThrow(StockDBOpenHelper.HistoricalDates._UPDATED)));
            }
            c.close();
        }

        return result;
    }

    @Override
    public void deleteAll() {
        deleteAll(StockDBOpenHelper.Prices.getTableName(Interval.DAILY));
        deleteAll(StockDBOpenHelper.Prices.getTableName(Interval.WEEKLY));
        deleteAll(StockDBOpenHelper.Prices.getTableName(Interval.MONTHLY));
        deleteAll(StockDBOpenHelper.HistoricalDates.TABLE_HISTORICAL_DATES_DAILY);
        deleteAll(StockDBOpenHelper.HistoricalDates.TABLE_HISTORICAL_DATES_WEEKLY);
        deleteAll(StockDBOpenHelper.HistoricalDates.TABLE_HISTORICAL_DATES_MONTHLY);

        optimize();
    }

    /*
    private void updatePrices(String symbolDescription, Interval interval) throws Exception {
        PriceList list = null;
        try
        {
            Calendar cal = Calendar.getInstance();
            cal.set(1990, Calendar.JANUARY, 1);
            //PriceList list = YahooFinance.getPrices(symbolDescription, interval, cal);
            //TODO limit for now
            list = mYahooFinance.getPrices(symbolDescription, interval, 500);
        }
        catch(Exception e)
        {
            // nothing
        }

        if (list != null && list.size() > 0) {
            add(list);
            Log.d(TAG, "Updated prices for " + symbolDescription);
            //mDb.log();
        }
        else {
            throw new Exception("Failed to get updated prices for " + symbolDescription);
        }
    }
    */
}

File: app/src/main/java/org/cerion/stockcharts/repository/PriceListSQLRepository.kt
package org.cerion.stockcharts.repository

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.cerion.stockcharts.common.TAG
import org.cerion.stockcharts.database.PriceListEntity
import org.cerion.stockcharts.database.PriceRowEntity
import org.cerion.stockcharts.database.getDatabase
import org.cerion.stocks.core.PriceList
import org.cerion.stocks.core.PriceRow
import org.cerion.stocks.core.model.FetchInterval
import org.cerion.stocks.core.platform.KMPDate
import org.cerion.stocks.core.repository.IPriceListRepository
import java.util.*

class PriceListSQLRepository(private val context: Context) : IPriceListRepository {

    private val roomDb = getDatabase(context)
    private val priceListDao = roomDb.priceListDao
    private val pricesDao = roomDb.pricesDao

    override fun get(symbol: String, interval: FetchInterval): PriceList? {
        val header = priceListDao.get(symbol, interval.ordinal)
                ?: return null

        val dbPrices = pricesDao.getAll(symbol, interval.ordinal)

        val prices = dbPrices.map {
            PriceRow(KMPDate(it.date), it.open, it.high, it.low, it.close, it.volume)
        }

        val result = PriceList(symbol, prices)
        result.lastUpdated = header.lastUpdated

        return result
    }

    override fun add(list: PriceList) {
        val interval = list.interval.ordinal

        // Delete existing + cascade deletes prices
        priceListDao.delete(list.symbol, interval)

        roomDb.runInTransaction {
            // Add list
            val priceList = PriceListEntity(list.symbol, interval)
            priceList.lastUpdated = Date()
            priceListDao.insert(priceList)

            // Add prices
            val dbPrices = list.map {
                PriceRowEntity(list.symbol, interval, it.date.jvmDate, it.open, it.high, it.low, it.close, it.volume)
            }

            pricesDao.insert(dbPrices)
        }
    }

    suspend fun clearCache() {
        withContext(Dispatchers.IO) {
            val name = roomDb.openHelper.databaseName
            val file = context.getDatabasePath(name)

            Log.i("Main", "Size before compact: ${file.length()}")
            priceListDao.deleteAll()
            pricesDao.deleteAll()
            roomDb.compact() // TODO how to verify this worked and how much space was saved

            // There is a delay for when this gets reported correctly, ignore for now
            //Log.i("Main", "Size after compact: ${file.length()}")
        }
    }

    suspend fun cleanupCache() {
        withContext(Dispatchers.IO) {
            val name = roomDb.openHelper.databaseName
            val file = context.getDatabasePath(name)
            Log.i("Main", "Size before compact: ${file.length()}")

            val lists = priceListDao.getAll().sortedBy { it.lastUpdated }
            val maxToKeep = 20
            val minToKeep = 5
            if (lists.size > maxToKeep) {
                val delete = lists.subList(0, lists.size - minToKeep)
                Log.i(TAG, "Deleting ${delete.size} cached lists")
                delete.forEach {
                    pricesDao.delete(it.symbol, it.interval)
                    priceListDao.delete(it)
                }

                roomDb.compact()
            }
        }
    }

    /*
    override fun deleteAll() {
        throw RuntimeException("not implemented")
        deleteAll(StockDBOpenHelper.Prices.getTableName(Interval.DAILY))
        deleteAll(StockDBOpenHelper.Prices.getTableName(Interval.WEEKLY))
        deleteAll(StockDBOpenHelper.Prices.getTableName(Interval.MONTHLY))
        deleteAll(StockDBOpenHelper.HistoricalDates.TABLE_HISTORICAL_DATES_DAILY)
        deleteAll(StockDBOpenHelper.HistoricalDates.TABLE_HISTORICAL_DATES_WEEKLY)
        deleteAll(StockDBOpenHelper.HistoricalDates.TABLE_HISTORICAL_DATES_MONTHLY)
        optimize()
    }
     */
}